src/machine/state/mod.rs

```diff
@@ -16,6 +17,49 @@ pub struct State {
 
 impl State {
     pub fn new() -> Self {
-        unimplemented!();
+        State {
+            // for the moment start execution at address 0
+            pc: 0,
+
+            sr: SRegisters::new(),
+            gpr: GPRegisters::new(),
+
+            // for the moment create 1MB of memory
+            mem: Memory::with_capacity(0x100000)
+        }
+    }
+
+    /// Display the state partially on the command line for testing purposes
+    pub fn debug_output(&self) {
```

Comment:
My first problem is: At the moment you specify a public function (which is therefore an API function) that does exactly the same thing as a Display trait. Hence - as long as you do not specify the parameters - you can write a Display trait implementation.

My second problem: You write to stdout from a function deep within the machine library. Library function should not have side-effect like this.

My actual idea is:
1) The module `machine` provides VERY BASIC Display and Debug traits.
2) For anything else, we write another module (called `io` or `view`) which implements a special `View` type.
3) A `View` is a struct that hold a reference to a machine and a lot of parameters that specify the view into the machine. Output is - again - done via a Display trait implementation.